File: January/10_Word_Subsets/ShaFeiii.cpp
#include <bits/stdc++.h>
using namespace std;


// c++ Solution
class Solution {
    vector<int> getStringFrequency(string& word) {
        vector<int> freq(26, 0);
        for (char word2 : word) {
            freq[word2 - 'a']++;
        }
        return freq;
    }
    bool compare(vector<int> &original, string& s) {
        vector<int> freq2 = getStringFrequency(s);
        for (int i = 0; i < 26; ++i) {
            if (freq2[i] < original[i]) return false;
        }
        return true;
    }
public:
    vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
        vector<string> ans;
        vector<int> original(26, 0);
        for (string& word : words2) {
            vector<int> temp = getStringFrequency(word);
            for (int i = 0; i < 26; ++i) {
                original[i] = max(original[i], temp[i]);
            }
        }
        for (string word : words1) {
            if (compare(original, word)) {
                ans.push_back(word);
            }
        }
        return ans;
    }
};

int main() {

    return 0;
}

File: January/10_Word_Subsets/ShaFeiii.py
class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        freq = [0] * 26
        for word in words2:
            temp = self.countFreq(word)
            for i in range(26):
                freq[i] = max(freq[i], temp[i])

        universal = []
        for word in words1:
            if self.compare(freq, word):
                universal.append(word)

        return universal

    def countFreq(self, word):
        freq = [0] * 26
        for w in word:
            freq[ord(w) - ord('a')] += 1
        return freq

    def compare(self, freq, word):
        freq2 = self.countFreq(word)
        for i in range(26):
            if freq[i] > freq2[i]:
                return False
        return True